Avoid allocations in Display for OsStr and Path

stepancheg opened the pull request:
#38879

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented on src/libstd/path.rs:
```diff
@@ -2281,8 +2281,8 @@ impl AsRef<OsStr> for Path {
-    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        self.inner.fmt(formatter)
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(&self.inner, formatter)
```
is this just general cleanup, or was this required for some reason?

stepancheg replied:
I've made several similar changes for these reasons:

* it is easier to visually check that correct trait is called (`Debug` or `Display`)
* to make sure code on Windows and Redox (e. g. `os_str::Slice`) works correctly. I do not have access to Windows machine, so it's easier to write code uniformly (single `fmt` import and explicit trait names) than wait for build system job completion
* and after all to unify all implementations in stack `os_str`, `OsStr`, `Path`

nikomatsakis commented on src/libstd_unicode/lossy.rs:
```diff
@@ -0,0 +1,154 @@
+
+/// Iterator over lossy UTF-8 string
+#[unstable(feature = "str_internals", issue = "0")]
+pub struct Utf8LossyIter<'a> {
```
can we write some more unit tests for this? it seems like quite a big chunk of code, and one for which it'd be easy to write some targeted unit tests...

arthurprs commented on src/libcollections/string.rs:
```diff
@@ -535,111 +535,37 @@ impl String {
 
-        if i > 0 {
-            unsafe { res.as_mut_vec().extend_from_slice(&v[..i]) };
+        let mut res = String::with_capacity(v.len());
```
We know it'll be larget than v.len right? Maybe reserve some extra, maybe v.len() * 3 / 2?

stepancheg replied:
Probably. If yes, this should be done in separate PR. I've just kept this part as it was before my patch.

stepancheg commented:
I've added tests (copied from `String::from_utf8_lossy` tests).

stepancheg commented:
Some tests broken, wait a bit...

stepancheg commented:
There's a tidy check error:

```
tidy error: /Users/yozh/devel/left/rust/src/libstd/ffi/os_str.rs:321: different `since` than before
tidy error: /Users/yozh/devel/left/rust/src/libstd/ffi/os_str.rs:685: different `since` than before
```

There were no `Display` for `OsStr` and `OsString` which are need to implement `Display` for `Path` and which I've added. What should I do?

I've added

```
#[stable(feature = "rust1", since = "1.20.0")]
```

which is apparently incorrect.

nikomatsakis commented:
Going to pass the buck here. I'm not that familiar with the attributes that need to be set.

r? @alexcrichton

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -312,12 +312,19 @@ impl Default for OsString {
 }
 
+#[stable(feature = "rust1", since = "1.20.0")]
+impl fmt::Display for OsString {
```
This did not exist before and is a new stable feature being added in this PR, is that intentional? It was the libs team intention that this impl did *not* exist.

stepancheg replied:
`Display` for `OsString` seems to be perfectly reasonable for me. You may need to "display" an OS string like you "display" the usual string.

Anywhay, if it should exist, how should I mute `tidy`?

If it shouldn't exist, how should I keep an implementation, but hide it from public API? `#[doc(hidden)]` and `#[unstable(feature = "str_internals", issue = "0")]`?

alexcrichton replied:
Whether or not it's reasonable it is intentionally excluded and should not be added in this PR. If you'd like to add it then that should be done through a separate PR or an RFC.

This shouldn't exist, which means it should be deleted (along with most other `Display` impls added here)

alexcrichton commented on src/libcore/fmt/mod.rs:
```diff
@@ -1512,23 +1512,30 @@ impl Display for bool {
 
+#[doc(hidden)]
+#[unstable(feature = "str_internals", issue = "0")]
+pub fn escape_str(s: &str, f: &mut Formatter) -> Result {
```
Can this avoid being exposed? It seems like a reasonable amount that's not too bad to just copy directly into the new iterator.

bors commented:
:umbrella: The latest upstream changes (presumably #42648) made this pull request unmergeable. Please resolve the merge conflicts.

stepancheg commented:
Updated the PR.

* Removed public `Display` implementations (kept implementations for private types in `os_str`)
* Copied implementation of `Debug` for `str` to `Debug` for `Utf8Lossy` instead of extraction of `escape_str` function

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit ac96fd7 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton ac96fd77874bb7968b7d82bed5a410d3941bc4c8 -->

bors commented:
:hourglass: Testing commit ac96fd7 with merge 3438c0f...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3609), [status-travis](https://travis-ci.org/rust-lang/rust/builds/243891477?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 3438c0fa8c45255ac7b8813c1253d1b59febb081 to master...
